### ISSUE TITLE
consul_prepared_query always returns unclean diff with `failover.nearest_n` set to 0 and `failover.datacenters` populated

### ISSUE THREAD

**Issue by aglover-zendesk** [terraform-providers/terraform-provider-consul#119]
### Terraform Version

```
$ terraform -v
Terraform v0.11.13
+ provider.consul v2.4.0
```

### Affected Resource(s)
- consul_prepared_query

### Terraform Configuration Files
```hcl
resource "consul_prepared_query" "use1" {
  datacenter   = "use1"
  name         = "${var.name}"
  service      = "${var.service}"
  tags         = "${var.tags}"
  only_passing = "${var.only_passing}"
  near         = "${var.near}"

  dns {
    ttl = "${var.dns_ttl}"
  }

  failover {
    nearest_n   = "0"
    datacenters = ["use1", "usw2"]
  }

  template {
    type   = "${var.template_type}"
    regexp = "${var.regexp}"
  }
}
```

### Expected Behavior
A `terraform` plan shows a clean diff with the message `No changes. Infrastructure is up-to-date.` after these resources are created/updated.

### Actual Behavior
Even after applying changes numerous times, `terraform plan` always shows that the failover properties are not set correctly, like this:
```
  ~ module.greensql-accountdb-prepared-query.consul_prepared_query.use1
      failover.#:               "0" => "1"
      failover.0.datacenters.#: "0" => "2"
      failover.0.datacenters.0: "" => "use1"
      failover.0.datacenters.1: "" => "usw2"
      failover.0.nearest_n:     "" => "0"
```

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. Create a new `consul_prepared_query` resource with at least one value for `failover.datacenters` and with `failover.nearest_n` set to 0
2. Execute `terraform apply`
3. Execute `terraform plan` - the prepared queries will show up as requiring modification

### Important Factoids
This behavior started occurring after upgrading from 2.3 to 2.4 of this provider. I believe this is the specific code change that caused the unwanted change in behavior - https://github.com/terraform-providers/terraform-provider-consul/pull/109/files#diff-ababb94278605280d50a21e47d80f840L198

### References
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:
- GH-109, specifically https://github.com/terraform-providers/terraform-provider-consul/pull/109/files#diff-ababb94278605280d50a21e47d80f840L198

**Comment by remilapeyre** [comment]
Thanks for reporting this bug @aglover-zendesk. The previous behavior of `consul_prepared_query` was not setting `failover` correctly so it would ignore external changes to this attribute. #109 fixes that but introduces this bug since I kept the old test without updating it. Changing the test `pq.Service.Failover.NearestN > 0` to `pq.Service.Failover.NearestN > 0 || len(pq.Service.Failover.Datacenters) > 0` fix the bug but there is still another issue. Since `failover` is a list, it does not consider this:

```
resource "consul_prepared_query" "use1" {
  datacenter   = "dc1"
  name         = "test"
  service      = "test"
  tags         = ["test"]
  only_passing = false
  near         = "dc1"

  dns {
    ttl = "10s"
  }

  failover {
    nearest_n   = "0"
    datacenters = []
  }

  template {
    type   = "name_prefix_match"
    regexp = "^(.*)-near-self$"
  }
}
```

and this:

```
resource "consul_prepared_query" "use1" {
  datacenter   = "dc1"
  name         = "test"
  service      = "test"
  tags         = ["test"]
  only_passing = false
  near         = "dc1"

  dns {
    ttl = "10s"
  }

  template {
    type   = "name_prefix_match"
    regexp = "^(.*)-near-self$"
  }
}
```

to be the same thing.


While not a huge issue when writing the configuration by hand, it is a problem when using variables:

```
resource "consul_prepared_query" "use1" {
  datacenter   = "dc1"
  name         = "test"
  service      = "test"
  tags         = ["test"]
  only_passing = false
  near         = "dc1"

  dns {
    ttl = "10s"
  }

  failover {
    nearest_n   = "${var.zero}"
    datacenters = "${var.empty_list}"
  }

  template {
    type   = "name_prefix_match"
    regexp = "^(.*)-near-self$"
  }
}
```

I think fixing this issue will require to change the type of `failover` but this would be a major change. This was not a problem before as it was hidden by the fact that `failover` was not updated correctly.

I think we could proceed in two changes here, first fix the issue you reported in a minor release, then change the type of `failover` (and probably the other attributes too) in a major release. Since it's a major change I would like the input of @pearkes before proceeding though.

**Comment by aglover-zendesk** [comment]
Thanks for the response @remilapeyre, and you inferred my problem perfectly - as you guessed, I'm setting `nearest_n` and `datacenters` with variables. I just used the explicit values in the issue report above for clarity.

Thanks for the detailed explanation of the behavior of the `failover` properties. A minor release that includes your fixes from #109 and fixes this diff behavior would be awesome!
